packetbeat/protos/tcp/tcp.go

```diff
@@ -75,117 +75,136 @@ type TcpStream struct {
-			if isDebug {
-				debugf("Stream doesn't exist, creating new")
-			}
-
-			// create
-			stream = &TcpStream{id: tcp.getId(), tuple: &pkt.Tuple, protocol: protocol, tcp: tcp}
-			stream.tcptuple = common.TcpTupleFromIpPort(stream.tuple, stream.id)
-			tcp.streams.PutWithTimeout(pkt.Tuple.Hashable(), stream, timeout)
-			created = true
-		} else {
-			original_dir = TcpDirectionReverse
-		}
+	stream, created := tcp.getStream(pkt)
+	if stream.conn == nil {
+		return
 	}
+	conn := stream.conn
+
 	tcp_start_seq := tcphdr.Seq
 	tcp_seq := tcp_start_seq + uint32(len(pkt.Payload))
-
+	lastSeq := conn.lastSeq[stream.dir]
 	if isDebug {
 		debugf("pkt.start_seq=%v pkt.last_seq=%v stream.last_seq=%v (len=%d)",
-			tcp_start_seq, tcp_seq, stream.lastSeq[original_dir], len(pkt.Payload))
+			tcp_start_seq, tcp_seq, lastSeq, len(pkt.Payload))
 	}
 
-	if len(pkt.Payload) > 0 &&
-		stream.lastSeq[original_dir] != 0 {
-
-		if tcpSeqBeforeEq(tcp_seq, stream.lastSeq[original_dir]) {
+	if len(pkt.Payload) > 0 && lastSeq != 0 {
+		if tcpSeqBeforeEq(tcp_seq, lastSeq) {
 			if isDebug {
-				debugf("Ignoring what looks like a retransmitted segment. pkt.seq=%v len=%v stream.seq=%v",
-					tcphdr.Seq, len(pkt.Payload), stream.lastSeq[original_dir])
+				debugf("Ignoring retransmitted segment. pkt.seq=%v len=%v stream.seq=%v",
+					tcphdr.Seq, len(pkt.Payload), lastSeq)
 			}
 			return
 		}
 
-		if tcpSeqBefore(stream.lastSeq[original_dir], tcp_start_seq) {
+		if tcpSeqBefore(lastSeq, tcp_start_seq) {
 			if !created {
-				logp.Warn("Gap in tcp stream. last_seq: %d, seq: %d", stream.lastSeq[original_dir], tcp_start_seq)
-				drop := stream.gapInStream(original_dir,
-					int(tcp_start_seq-stream.lastSeq[original_dir]))
+				gap := int(tcp_start_seq - lastSeq)
+				logp.Warn("Gap in tcp stream. last_seq: %d, seq: %d, gap: %d", lastSeq, tcp_start_seq, gap)
+				drop := stream.gapInStream(gap)
 				if drop {
 					if isDebug {
-						debugf("Dropping stream because of gap")
+						debugf("Dropping connection state because of gap")
 					}
-					tcp.streams.Delete(stream.tuple.Hashable())
+
+					// drop application layer connection state and
+					// update stream_id for app layer analysers using stream_id for lookups
+					conn.id = tcp.getId()
+					conn.data = nil
 				}
 			}
 		}
 	}
-	stream.lastSeq[original_dir] = tcp_seq
 
-	stream.addPacket(pkt, tcphdr, original_dir)
+	conn.lastSeq[stream.dir] = tcp_seq
+	stream.addPacket(pkt, tcphdr)
+}
+
+func (tcp *Tcp) getStream(pkt *protos.Packet) (stream TcpStream, created bool) {
+	if conn := tcp.findStream(pkt.Tuple.Hashable()); conn != nil {
+		return TcpStream{conn: conn, dir: TcpDirectionOriginal}, false
+	}
+
+	if conn := tcp.findStream(pkt.Tuple.RevHashable()); conn != nil {
+		return TcpStream{conn: conn, dir: TcpDirectionReverse}, false
+	}
+
+	protocol := tcp.decideProtocol(&pkt.Tuple)
+	if protocol == protos.UnknownProtocol {
+		// don't follow
+		return TcpStream{}, false
+	}
+
+	timeout := time.Duration(0)
```

Comment:
ah, copy'n paste. I'd prefer the 'var timeout time.Duration'. but in this case no big difference